p2p/net/swarm/swarm_dial.go

```diff
@@ -468,6 +468,8 @@ func (s *Swarm) filterKnownUndialables(p peer.ID, addrs []ma.Multiaddr) []ma.Mul
 		func(addr ma.Multiaddr) bool {
 			return s.gater == nil || s.gater.InterceptAddrDial(p, addr)
 		},
+		s.udpBHD.IsDialAllowed,
+		s.ipv6BHD.IsDialAllowed,
```

Comment:
Would be nice to have just a single black hole detector. This detector would then hand off to one IPv6 and UDP blackhole detector.

Reply:
To clarify:
```go
type blackHoleDetector struct {
    udp, ipv6 singleBlackHoleDetector // TODO: better naming
}

func (d *blackHoleDetector) RecordResult(addr ma.Multiaddr, result bool) {
    if d.udp != nil & isUDPAddr(addr) {
         d.udp.RecordResult(addr, result)
    }
    if d.ipv6 != nil & isIPv6Addr(addr) {
         d.ipv6.RecordResult(addr, result)
    }
}